--- gcc-4.3.2-1/gcc/Makefile.in	2009-12-21 18:17:41.000000000 +0200
+++ gcc-4.3.2-2/gcc/Makefile.in	2009-12-21 18:17:25.000000000 +0200
@@ -377,7 +377,7 @@
 MD5_H	    = $(srcdir)/../include/md5.h
 
 # Default native SYSTEM_HEADER_DIR, to be overridden by targets.
-NATIVE_SYSTEM_HEADER_DIR = /usr/include
+NATIVE_SYSTEM_HEADER_DIR = /include
 # Default cross SYSTEM_HEADER_DIR, to be overridden by targets.
 CROSS_SYSTEM_HEADER_DIR = @CROSS_SYSTEM_HEADER_DIR@
 
@@ -1633,7 +1633,8 @@
 	echo CRTSTUFF_T_CFLAGS = '$(CRTSTUFF_T_CFLAGS)' >> tmp-libgcc.mvars
 	echo CRTSTUFF_T_CFLAGS_S = '$(CRTSTUFF_T_CFLAGS_S)' >> tmp-libgcc.mvars
 
-	mv tmp-libgcc.mvars libgcc.mvars
+	cat tmp-libgcc.mvars | sed -e 's|-lc||g' > libgcc.mvars
+	rm tmp-libgcc.mvars
 
 # Use the genmultilib shell script to generate the information the gcc
 # driver program needs to select the library directory based on the
