diff -uNr u-boot-2010.03/board/owen/plc240/owen-plc240.c u-boot-2010.03-new/board/owen/plc240/owen-plc240.c
--- u-boot-2010.03/board/owen/plc240/owen-plc240.c	2013-10-17 11:02:18.724073517 +0400
+++ u-boot-2010.03-new/board/owen/plc240/owen-plc240.c	2013-10-16 10:10:17.724082581 +0400
@@ -91,6 +91,9 @@
 
 	/* Enable NandFlash */
 	at91_set_pio_output(CONFIG_SYS_NAND_ENABLE_PIN, 1);
+	
+	//VBUS!!
+	// AT91_PIN_PE13, AT91_PIN_PE14
 }
 #endif
 
@@ -135,6 +138,7 @@
 	/* Re-enable pull-up */
 	writel(1 << 25, &pio->pioc.puer);
 	writel((1 << 25) | (1 <<26), &pio->pioe.puer);
+	writel((1 << 13) | (1 <<14), &pio->pioe.puer);
 
 	at91_macb_hw_init();
 }
diff -uNr u-boot-2010.03/cpu/arm926ejs/at91/at91sam9263_devices.c u-boot-2010.03-new/cpu/arm926ejs/at91/at91sam9263_devices.c
--- u-boot-2010.03/cpu/arm926ejs/at91/at91sam9263_devices.c	2010-04-01 01:54:39.000000000 +0400
+++ u-boot-2010.03-new/cpu/arm926ejs/at91/at91sam9263_devices.c	2013-10-16 10:10:09.732757639 +0400
@@ -195,8 +195,8 @@
 void at91_uhp_hw_init(void)
 {
 	/* Enable VBus on UHP ports */
-	at91_set_pio_output(AT91_PIO_PORTA, 21, 0);
-	at91_set_pio_output(AT91_PIO_PORTA, 24, 0);
+	at91_set_pio_output(AT91_PIO_PORTE, 13, 0);
+	at91_set_pio_output(AT91_PIO_PORTE, 14, 0);
 }
 #endif
